### PR TITLE
[build-tools] add new `eas` custom build function for app repacking

### PR DESCRIPTION
**Pull request by szdziedzic**
# Why

Repack onboarding builds

# How

Add new `eas/__download_and_repack_golden_development_client_archive` custom build function to use `@expo/repack-app` to do the repack + resign

# Test Plan

Create a new repo from scratch using the onboarding flow

Add this build profile to your `eas.json`
```json
"repack": {
      "developmentClient": true,
      "distribution": "internal",
      "channel": "development",
      "android": {
        "image": "latest"
      },
      "ios": {
        "image": "latest",
        "config": "repack.yml"
      }
},
```

Add `.eas/build/repack.yml`
```yml
build:
  name: Demo
  steps:
    - eas/checkout
    - eas/install_node_modules
    - eas/__download_and_repack_golden_development_client_archive:
        id: download_and_repack_golden_development_client_archive
    - eas/upload_artifact:
        name: Upload build artifact
        inputs:
          type: application-archive
          path: ${ steps.download_and_repack_golden_development_client_archive.repacked_archive_path }
```

Run `eas build -p ios -e repack`

Screenshot 2024-05-16 at 10 42 29

### CODE REVIEW

File: packages/build-tools/src/utils/artifacts.ts [hunk 1]
```diff
@@ -19,7 +19,11 @@ export async function findArtifacts({
   /** If provided, will log error suggesting possible files to upload. */
   logger: bunyan | null;
 }): Promise<string[]> {
-  const files = await fg(patternOrPath, { cwd: rootDir, onlyFiles: false });
+  const files = path.isAbsolute(patternOrPath)
```
Comment (sjchmiela) [inline packages/build-tools/src/utils/artifacts.ts:22]: Hmm, doesn't `fg` support absolute paths? That's what I thought…

File: packages/build-tools/src/steps/functions/repack.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    id: '__download_and_repack_golden_development_client_archive',
+    name: 'Download and repack golden development client archive',
+    outputProviders: [
+      BuildStepOutput.createProvider({
+        id: 'repacked_archive_path',
+        required: true,
+      }),
+    ],
+    fn: async (stepsCtx, { outputs, env }) => {
+      const tmpDir = `/tmp/eas-build-golden-dev-client-app-${uuidv4()}`;
```
Comment (sjchmiela) [inline packages/build-tools/src/steps/functions/repack.ts:29]: Shouldn't we use `os.tmpdir()`?

File: packages/build-tools/src/steps/functions/repack.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      stepsCtx.logger.info(`Created temporary directory: ${tmpDir}`);
+
+      stepsCtx.logger.info('Downloading golden development client archive...');
+      const fileName =
+        stepsCtx.global.staticContext.job.platform === Platform.IOS
+          ? 'golden-dev-client-latest.ipa'
+          : 'golden-dev-client-latest.apk';
+      const goldenArchiveUrl =
+        env.__EAS_GOLDEN_DEV_CLIENT_URL ??
+        `https://storage.googleapis.com/turtle-v2/onboarding/${fileName}`;
```
Comment (sjchmiela) [inline packages/build-tools/src/steps/functions/repack.ts:40]: Maybe we can put it under `turtle-v2-artifacts` and use `job-artifacts.eascdn.net` to reduce CF costs? Or maybe we should add an `eascdn.net` subdomain for `turtle-v2`?

File: packages/build-tools/src/steps/functions/repack.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      const goldenArchivePath = path.join(tmpDir, fileName);
+      try {
+        const response = await fetch(goldenArchiveUrl, {
+          timeout: 5 * 60 * 1000, // 5 minutes
```
Comment (sjchmiela) [inline packages/build-tools/src/steps/functions/repack.ts:44]: Why so huge? How big are the archives? Maybe we should have lesser timeout and retry?

File: packages/build-tools/src/steps/functions/repack.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+          outputPath: repackedArchivePath,
+          workingDirectory: tmpDir,
+          androidSigningOptions: androidCredentials
+            ? {
+                keyStorePath: androidCredentials?.keyStorePath,
+                keyStorePassword: androidCredentials?.keyStorePassword,
+                keyAlias: androidCredentials?.keyAlias,
+                keyPassword: androidCredentials?.keyPassword,
+              }
+            : undefined,
```
Comment (sjchmiela) [inline packages/build-tools/src/steps/functions/repack.ts:126]:
```suggestion
          androidSigningOptions: androidCredentials,
```
hm?

File: packages/build-tools/resources/__eas/repack.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  steps:
+    - eas/checkout
+    - eas/install_node_modules
+    - eas/__download_and_repack_golden_development_client_archive:
+        id: download_and_repack_golden_development_client_archive
+    - eas/upload_artifact:
+        name: Upload build artifact
+        inputs:
+          type: application-archive
+          path: ${ steps.download_and_repack_golden_development_client_archive.repacked_archive_path }
```
Comment (sjchmiela) [inline packages/build-tools/resources/__eas/repack.yml]:
```suggestion
          path: ${ steps.download_and_repack_golden_development_client_archive.repacked_archive_path }

```
newline at the end of file

Comment (szdziedzic) [inline packages/build-tools/resources/__eas/repack.yml:12]:
```suggestion
          path: ${ steps.download_and_repack_golden_development_client_archive.repacked_archive_path }
          
```

File: packages/build-tools/resources/__eas/repack.yml [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
```
Comment (sjchmiela) [inline packages/build-tools/resources/__eas/repack.yml:13]:
```suggestion

```

File: packages/build-tools/resources/__eas/repack.yml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+build:
+  name: Repack golden app
```
Comment (sjchmiela) [inline packages/build-tools/resources/__eas/repack.yml:2]:
```suggestion
  name: Resign development client
```
more user friendly?

**Review by sjchmiela — APPROVED** [review]

File: packages/build-tools/src/steps/functions/repack.ts (hunk 2 above)
Comment (szdziedzic) [reply packages/build-tools/src/steps/functions/repack.ts:29]: we should!

File: packages/build-tools/src/steps/functions/repack.ts (hunk 5 above)
Comment (szdziedzic) [reply packages/build-tools/src/steps/functions/repack.ts:126]: lol, thanks

File: packages/build-tools/src/steps/functions/repack.ts (hunk 3 above)
Comment (szdziedzic) [reply packages/build-tools/src/steps/functions/repack.ts:40]: I will try to do it in separate PR

File: packages/build-tools/src/utils/artifacts.ts (hunk 1 above)
Comment (szdziedzic) [reply packages/build-tools/src/utils/artifacts.ts:22]: It failed to upload `/tmp/...` for me 🤔

File: packages/build-tools/src/steps/functions/repack.ts (hunk 4 above)
Comment (szdziedzic) [reply packages/build-tools/src/steps/functions/repack.ts]: I made it 1 min

File: packages/build-tools/src/utils/artifacts.ts (hunk 1 above)
Comment (Kudo) [reply packages/build-tools/src/utils/artifacts.ts:22]: if that's macos, maybe you could try to use [temp-dir](https://www.npmjs.com/package/temp-dir) than `os.tmpdir()`

File: packages/build-tools/src/steps/functions/repack.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    ],
+    fn: async (stepsCtx, { outputs, env }) => {
+      const tmpDir = path.join(os.tmpdir(), `eas-build-golden-dev-client-app-${uuidv4()}`);
+      await fs.mkdirs(tmpDir);
+      stepsCtx.logger.info(`Created temporary directory: ${tmpDir}`);
+
+      stepsCtx.logger.info('Downloading golden development client archive...');
+      const fileName =
+        stepsCtx.global.staticContext.job.platform === Platform.IOS
+          ? 'golden-dev-client-latest.ipa'
```
Comment (Kudo) [inline packages/build-tools/src/steps/functions/repack.ts:37]: question: do we need to have both iphoneos and iphonesimulator golden ipa?

File: packages/build-tools/src/steps/functions/repack.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        );
+        const credentials = await credentialsManager.prepare(stepsCtx.logger);
+        await repackAppIosAsync({
+          platform: 'ios',
+          projectRoot: stepsCtx.workingDirectory,
+          sourceAppPath: goldenArchivePath,
+          outputPath: repackedArchivePath,
+          workingDirectory: tmpDir,
+          iosSigningOptions: {
+            provisioningProfile: Object.values(credentials.targetProvisioningProfiles)[0].path,
```
Comment (Kudo) [inline packages/build-tools/src/steps/functions/repack.ts:77]: i could update repack-app to support provisioningProfile as an array later

File: packages/build-tools/src/steps/functions/repack.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+          skipWorkingDirCleanup: true,
+          verbose: env.__EAS_REPACK_VERBOSE !== undefined,
+        });
+      } else if (stepsCtx.global.staticContext.job.platform === Platform.ANDROID) {
+        let androidCredentials:
+          | {
+              keyStorePath: string;
+              keyStorePassword: string;
+              keyAlias: string;
+              keyPassword: string | undefined;
```
Comment (Kudo) [inline packages/build-tools/src/steps/functions/repack.ts:92]: i can export the `AndroidSigningOptions` type from repack-app later

**Review by Kudo — APPROVED** [review]
the only concern is the iphoneos / iphonesimulator golden ipa. do we like to support this?

**Pull request merged**